### ISSUE TITLE
Inconsistent focus on new window

### ISSUE THREAD

**Issue by TheTerrior** [qtile/qtile#3818]
### The issue:

Version: 0.21.0-1
Log: no relevant logs

When I open a new window, whether the current window keeps focus or the new window takes focus feels completely random. To replicate, I have one window open on my main monitor (I have two). Place the mouse on the left half of the screen and open a new window (should open on the right half). The focus will go to the new window half the time and it should stay on the current window half the time. My mouse has not moved, I'm simply opening and closing that new window every time, and the focus is random every time I open it again.

I've tried this:

    @hook.subscribe.client_new
    def client_new(client):
        client.cmd_focus()
        hook.fire("focus_change")

but this has seemingly changed nothing. Additionally, I've set follow_mouse_focus to True.

### Required:

- [X] I have searched past issues to see if this bug has already been reported.

**Comment by m-col** [comment]
Could you try with `@hook.subscribe.client_managed`?

**Comment by TheTerrior** [comment]
I replaced it, and it seems to have fixed it. I switched it back to what it was originally and now I can't replicate the original issue. Maybe some sort of update to my system fixed the problem.

**Comment by m-col** [comment]
OK, great. Feel free to re-open if you have issues again.

**Comment by TheTerrior** [comment]
I've begun to use qtile again and have run into this issue once again. I've discovered that the problem is happening with VSCode, but doesn't seem to happen with any other application. Here's a video on youtube demonstrating the issue: https://youtu.be/CqbKBHdFgq4

**Comment by m-col** [comment]
That's unfortunate. Just to confirm, this is in X11? Please could you paste debug logs (`qtile cmd-obj -o cmd -f debug`) for one time when the new window opens and the focus stays at the left, and again for when it opens and the focus changes to the new window?

Is your hook above necessary? Are you hooking anything to `focus_change`? I don't think you should be firing hooks manually, in fact i think `client.cmd_focus()` will fire `focus_change` itself, so it's getting fired twice.

You may be able to work around by making the hook async, because then it will fire only after the current task completes:
```py
@hook.subscribe.client_new
async def client_new(client):
    client.focus()
```
However in general I think new windows, if they open on the current group (or maybe any visible group? can't remember) will steal focus

**Comment by elParaguayo** [comment]
> You may be able to work around by making the hook async, because then it will fire only after the current task completes:

Is that what happens when they're async? We should make that clearer in the docs.

**Comment by m-col** [comment]
(Disclaimer that this is my current understanding, if I am mistaken corrections are encouraged!) Async hooks are passed to `asyncio.ensure_future()` (which TIL is deprecated from 3.10), which schedules the coroutine to run in the background. But 'run in the background' doesn't mean with true parallelism because we're not doing any multiprocessing. The current task (which fired the hook) has control and will continue to execute until either it finishes or an `await` is hit. At that point asyncio will move on to the next task, which could be the hook. We don't do much async stuff outside of some widgets (and obviously the event loop), so essentially async hooks are being queued and called 'soon'.

Of course, you shouldn't need to know all of that for the docs to be clear to you so yeah we could update it :)

**Comment by elParaguayo** [comment]
Nice. That was my interpretation too.

As for the `ensure_future` being deprecated, that's fixed in my MPRIS2 PR.

**Comment by TheTerrior** [comment]
Yes this is in X11, I haven't tried out Wayland as of now. With all due respect, what is the debug command doing? Is it generating a file somewhere that I need to find? When I run it, it just doesn't return anything in the terminal itself. I tried the async solution but it didn't change anything.

**Comment by m-col** [comment]
It enables debug-level logging. Qtile sends log messages to `~/.local/share/qtile/qtile.log`. Note debug logging is very verbose, so ideally when trying to get debug logs during an event you might try to identify the period of logs just for that period.

You could do this instead:
```py
from libqtile import qtile

@hook.subscribe.client_new
def client_new(client):
    qtile.debug()  # turn on debug logging
    client.focus()
    qtile.warning()  # turn off debug logging
```
Then when the log level changes it should be pretty clear in the logs and it will help us find out what's going on :)

**Comment by TheTerrior** [comment]
Okay, getting those logs was a journey but I got it, here's the results.
[stayed_logs.txt](https://github.com/qtile/qtile/files/9956006/stayed_logs.txt)
[switched_logs.txt](https://github.com/qtile/qtile/files/9956009/switched_logs.txt)
